File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/cluster/ClusterIdResolver.scala
package net.gree.aurora.scala.domain.cluster

import scala.language.implicitConversions
import net.gree.aurora.domain.cluster.{ClusterIdResolver => JClusterIdResolver}
import net.gree.aurora.domain.hint.{Hint => JHint}
import net.gree.aurora.scala.domain.hint.Hint
trait ClusterIdResolver[T] extends ((Hint[T], Int) => ClusterId) {

  def apply(value: Hint[T], clusterSize: Int): ClusterId

}


object ClusterIdResolver {

  def apply[T](f: JClusterIdResolver[T]): ClusterIdResolver[T] = {
    new ClusterIdResolver[T] {
      def apply(hint: Hint[T], clusterSize: Int): ClusterId = ClusterId(f.apply(hint, clusterSize))
    }
  }


  implicit def toJava[T](shardIdResolver: ClusterIdResolver[T]) = {
    new JClusterIdResolver[T] {
      def apply(hint: JHint[T], clusterSize: Int) = shardIdResolver(Hint(hint), clusterSize)
    }
  }

}

File: aurora-core/src/main/java/net/gree/aurora/domain/cluster/ClusterIdResolver.java
package net.gree.aurora.domain.cluster;

import net.gree.aurora.domain.hint.Hint;

/**
 * {@link ClusterId}をヒントから解決するためのインターフェイス。
 *
 * @param <T> ヒントの型
 */
public interface ClusterIdResolver<T> {

    /**
     * ヒントから{@link ClusterId}を解決する。
     *
     * @param hint ヒント
     * @return {@link ClusterId}
     */
    ClusterId apply(Hint<T> hint, int clusterSize);

}

